enterprise/backend/src/metabase_enterprise/workspaces/merge.clj

```diff
@@ -3,31 +3,52 @@
    All functions assumes that validation of the dependencies AFTER merge is done BEFORE calling any of these methods."
   (:require
    [metabase-enterprise.transforms.api :as transforms.api]
+   [metabase-enterprise.workspaces.models.workspace-merge]
+   [metabase-enterprise.workspaces.models.workspace-merge-transform]
    [metabase-enterprise.workspaces.types :as ws.types]
    [metabase.util.malli :as mu]
    [toucan2.core :as t2])
   (:import (java.sql Connection)))
 
 (set! *warn-on-reflection* true)
 
+(defn- create-merge-transform-history!
+  "Create a workspace_merge_transform record to track this merge operation.
+   Does nothing for :noop operations (archived transforms that don't exist globally)
+   or :delete operations (history would be CASCADE deleted with the transform anyway)."
+  [op merge-map]
+  (when (#{:create :update} op)
+    (t2/insert! :model/WorkspaceMergeTransform merge-map)))
+
 ;; TODO (crisptrutski 2025/12/10): When there are more entity types support, this should update those too.
 (mu/defn merge-transform! :- [:map
                               [:op [:enum :create :delete :noop :update]]
                               [:ref_id ::ws.types/ref-id]
                               [:global_id {:optional true} [:maybe ::ws.types/appdb-id]]
                               [:error {:optional true} :any]]
   "Make the given transform in the Changeset public, i.e. create or update the relevant model/Transform entities.
-   This should also clear it out from the Changset, as it no longer has any changes."
-  [{:keys [global_id ref_id archived_at] :as ws-transform}]
+   This should also clear it out from the Changset, as it no longer has any changes.
+
+   Creates a WorkspaceMerge and WorkspaceMergeTransform history record.
+   - workspace: The workspace map (must have :id and :name keys)
+   - workspace-merge-id: ID of parent WorkspaceMerge (nil to create a new one for single-transform merges)
+   - merging-user-id: user performing the merge
+   - commit-message: description of the merge"
+  [{:keys [global_id ref_id archived_at] :as ws-transform} workspace workspace-merge-id merging-user-id commit-message]
   ;; Problems this may run into:
   ;; It will recalculate and validate the dag greedily as it inserts each items, and this might fail of temporary conflicts.
   ;; Some of these conflicts could be solved by ordering things smartly, but in the general case (e.g. reversing a
   ;; chain) this will not be possible - we would need to defer all validation and calculation, which is a bit scary.
   ;; Let's just focus on the happy path for now though, as this is a fundamental problem and unrelated to the design.
 
-  ;; Assume validation has alr
-  ;; Ensure the hook
-  (let [{:keys [error] :as result}
+  (let [ws-merge-id    (or workspace-merge-id
+                           (t2/insert-returning-pk!
```

Comment:
I'm not keen on this overloading of the function to cases it where it only creates it in one case. I'd rather that the caller always takes care of creation.